src/main/scala/io/iohk/ethereum/consensus/ConsensusImpl.scala

```diff
@@ -28,226 +20,168 @@ import io.iohk.ethereum.ledger.BlockData
+    * branch.
+    * @param branch                  the new branch as a sorted list of blocks. It's parent must
+    *                                be in the current best branch
     * @param blockExecutionScheduler threadPool on which the execution should be run
     * @param blockchainConfig        blockchain configuration
     * @return One of:
-    *   - [[BlockImportedToTop]] - if the block was added as the new best block
-    *   - [[BlockEnqueued]]      - block is stored in the [[io.iohk.ethereum.ledger.BlockQueue]]
-    *   - [[ChainReorganised]]   - a better new branch was found causing chain reorganisation
-    *   - [[DuplicateBlock]]     - block already exists either in the main chain or in the queue
-    *   - [[BlockImportFailed]]  - block failed to execute (when importing to top or reorganising the chain)
+    *   - [[ExtendedCurrentBestBranch]] - if the branch was added on top of the current branch
+    *   - [[SelectedNewBestBranch]]     - if the chain was reorganized.
+    *   - [[KeptCurrentBestBranch]]     - if the branch was not considered as better than the current branch
+    *   - [[ConsensusError]]            - block failed to execute (when importing to top or reorganising the chain)
+    *   - [[ConsensusErrorDueToMissingNode]]  - block failed to execute (when importing to top or reorganising the chain)
     */
   override def evaluateBranch(
-      branch: Seq[Block]
-  )(implicit blockExecutionScheduler: Scheduler, blockchainConfig: BlockchainConfig): Task[BlockImportResult] = {
-    val block = branch.head
+      branch: NonEmptyList[Block]
+  )(implicit blockExecutionScheduler: Scheduler, blockchainConfig: BlockchainConfig): Task[ConsensusResult] =
     blockchainReader.getBestBlock() match {
       case Some(bestBlock) =>
         blockchain.getChainWeightByHash(bestBlock.header.hash) match {
-          case Some(weight) => handleBlockImport(block, bestBlock, weight)
+          case Some(weight) => handleBranchImport(branch, bestBlock, weight)
           case None         => returnNoTotalDifficulty(bestBlock)
         }
       case None => returnNoBestBlock()
     }
-  }
 
-  private def handleBlockImport(block: Block, bestBlock: Block, weight: ChainWeight)(implicit
+  private def handleBranchImport(
+      branch: NonEmptyList[Block],
+      currentBestBlock: Block,
+      currentBestBlockWeight: ChainWeight
+  )(implicit
       blockExecutionScheduler: Scheduler,
       blockchainConfig: BlockchainConfig
-  ): Task[BlockImportResult] = {
-    val importResult = if (isPossibleNewBestBlock(block.header, bestBlock.header)) {
-      importToTop(block, bestBlock, weight)
-    } else {
-      reorganiseOrEnqueue(block, bestBlock, weight)
+  ): Task[ConsensusResult] = {
+    val parentHash = branch.head.header.parentHash
+    val parentWeight = blockchain
+      .getChainWeightByHash(parentHash)
+      .getOrElse(throw new Error("Inconsistent database"))
```

Comment:
Can you please add more info to the error? Like block hash/number

Reply:
Actually, I should maybe change this. I did it at the beginning and forgot to rewrite it. Ideally, I would prefer to just throw when that kind of things happens, because it's not the Consensus job to handle inconsistencies that should not happen in the first place. But currently the way of doing is returning a type to represent the error, so I should do that to stay compatible.

Reply:
I rewrote the method and added the hash of the block and its number in the message.

Reply:
Easier to read now 👍